### ISSUE TITLE
Feature Request: Generic IP Camera Add-On

### ISSUE THREAD

**Issue by dschmidt2020** [mozilla-iot/gateway#712]
Having already proposed this idea to the development team at Mozilla, I thought it would be best to throw it out there to the PT community.

I have used Home Assistant for a while now and despite it's flexibility to integrate a ton of devices, it's gateway setup and needing to edit .yaml files to configure simple backend settings was ridiculous.  Enter Project Things.

What we need is to build an add-on based on the [Onvif Profile S specification](https://www.onvif.org/profiles/profile-s/) that will allow a user to connect almost any ip camera to their raspberry pi and be able to access it through the IOT gateway.

HA has a bunch of components for IP Cameras (the FFMPEG setup via rtsp through opencv works well)

As far as proposed technologies it seems to me that using the [WebRTC extension for the UV4L Streaming Server](https://www.linux-projects.org/uv4l/webrtc-extension/) will be the best option as far as latency, security, scalability, ease of use, and cross browser compatibility.

I think it will integrate well in the PT web app.

Any thoughts???

**Comment by Galatoni** [comment]
I've recently picked up a pi 3 and got project things installed for just this reason. I was perhaps going to work on my own node implementation of integrating a webcam with the pi and casting to my TV via chrome cast. Libraries exist to do most of this already, which is brilliant. But would be great to have this functionality built in.

I've no idea where to start though and I've only ever worked directly for a company before. Never on open source projects. Do we have anything in the way of documentation for the API or are we not there yet?

**Comment by benfrancis** [comment]
> Do we have anything in the way of documentation for the API or are we not there yet?
* [Web Thing API](https://iot.mozilla.org/wot/) (exposed by the gateway)
* [Adapter IPC messages](https://github.com/mozilla-iot/wiki/wiki/Adapter-IPC) (how add-ons talk to the gateway)
* [Getting Started blog posts](https://hacks.mozilla.org/category/web-of-things/)

I'd be interested to know how Chromecast fits in, as that's not immediately obvious to me. Chromecast is basically a full screen web browser that you cast URLs to. See also, [W3C Presentation API](https://www.w3.org/TR/presentation-api/) which is the standards-based alternative/complement to Chromecast on the client side.

**Comment by Galatoni** [comment]
Well I'm certainly no expert. But it would be nice to be able to feed generic webcam feeds into the Chromecast. But yes, I think a more generic casting solution would be best. I see this idea as a means to an end.
Webcam feeds in - cast to a connected device. The Pi is super small so having it connected to a webcam, or having the gateway itself capable of receiving a feed would be a great idea.

Likely a two part process though. Issue for the webcam feed, and then another issue that sets up a casting socket? What do you think?

**Comment by dschmidt2020** [comment]
I echo that @Galatoni. I am no expert myself.  I can only go on past experience with getting HA setup.  I like the idea of somehow control wireless IP cameras, not usb webcams, with just Raspberry Pi. No chromecast. 

Getting familiar with WT API and Adapter IPC I'm kind of lost on the direction to head in.  I think rather than re-inventing the wheel there must be a way to integrate the bones of HA's open source [FFMPEG Component](https://home-assistant.io/components/camera.ffmpeg/) to build an add-on.

You would still need to compile ffmpeg from source and setup libavcodec from terminal on the pi.

**Comment by dschmidt2020** [comment]
This I think may be a better resource... [ONVIF Component](https://github.com/home-assistant/home-assistant/blob/master/homeassistant/components/camera/onvif.py)

**Comment by Galatoni** [comment]
Thanks so much for the resources @dschmidt2020 and @benfrancis 
I can't believe I couldn't find the API documentation myself 🤦‍♂️ 
I had a few ideas of how to achieve this in JavaScript already. I've been doing a fair bit with Websockets recently in Javascript, so it seemed a logical place to start for me. I hadn't considered a Python solution. I'll have a proper read of the API.

So with regards to that ONVIF stuff. That would take the feed and push it out as an FFmpeg stream, which we would then need to pipe through to our casting solution. Obviously we need map the controls to the UI and somehow pick a target. Quite a lot to think about here, even before we have to consider casting :D

**Comment by mrstegeman** [comment]
Add-on has been created, I'm just waiting on a schema at this point.

See: https://github.com/mozilla-iot/onvif-adapter

**Comment by cr0ybot** [comment]
Perhaps a separate add-on that requires both this ONVIF adapter and the existing Chromecast add-on by @freaktechnik would be possible? It would probably require additions to the Chromecast add-on to accept steams/URLs to cast. Now that I mention it, does a mechanism even exist for two separate add-ons to communicate with each other? Or for one add-on to "require" other add-ons?

**Comment by mrstegeman** [comment]
I like the idea, but there's currently no way for add-ons to interact with one another (unless they coordinate and do so outside of the normal API), nor is there a way for one add-on to require another.

**Comment by freaktechnik** [comment]
My solution would be to have a semantic action exposed on devices to play some media file or stream. There would have to be some way to know what protocols and formats exactly are supported etc. and we should check first how well that maps to Google Cast/Airplay/DLNA/Sonos/Smart assistants etc.
